database/pdb.ksy

```diff
@@ -0,0 +1,961 @@
+      - id: empty_candidate
+        type: u4
+      - id: first_page
+        type: page_ref
+        doc: |
+          Links to the chain of pages making up that table. The first
+          page seems to always contain similar garbage patterns and
+          zero rows, but the next page it links to contains the start
+          of the meaningful data rows.
+      - id: last_page
+        type: page_ref
+        doc: |
+          Holds the index of the last page that makes up this table.
+          When following the linked list of pages of the table, you
+          either need to stop when you reach this page, or when you
+          notice that the `next_page` link you followed took you to a
+          page of a different `type`.
+    -webide-representation: '{type}'
+
+  page_ref:
+    doc: |
+      An index which points to a table page (its offset can be found
+      by multiplying the index by the `page_len` value in the file
+      header). This type allows the linked page to be lazy loaded.
+    seq:
+      - id: index
+        type: u4
+        doc: |
+          Identifies the desired page number.
+    instances:
+      body:
+        doc: |
+          When referenced, loads the specified page and parses its
+          contents appropriately for the type of data it contains.
+        io: _root._io
+        pos: _root.len_page * index
+        size: _root.len_page
+        type: page
+
+  page:
+    doc: |
+      A table page, consisting of a short header describing the
+      content of the page and linking to the next page, followed by a
+      heap in which row data is found. At the end of the page there is
+      an index which locates all rows present in the heap via their
+      offsets past the end of the page header.
+    seq:
+      - contents: [0, 0, 0, 0]
+      - id: page_index
+        doc: Matches the index we used to look up the page, sanity check?
+        type: u4
+      - id: type
+        type: u4
+        enum: page_type
+        doc: |
+          Identifies the type of information stored in the rows of this page.
+      - id: next_page
+        doc: |
+          Index of the next page containing this type of rows. Points past
+          the end of the file if there are no more.
+        type: page_ref
+      - type: u4
+        doc: |
+          @flesniak said: "sequence number (0->1: 8->13, 1->2: 22, 2->3: 27)"
+      - size: 4
+      - id: num_rows_small
+        type: u1
+        doc: |
+          Holds the value used for `num_rows` (see below) unless
+          `num_rows_large` is larger (but not equal to `0x1fff`). This
+          seems like some strange mechanism to deal with the fact that
+          lots of tiny entries, such as are found in the
+          `playlist_entries` table, are too big to count with a single
+          byte. But why not just always use `num_rows_large`, then?
+      - type: u1
+        doc: |
+          @flesniak said: "a bitmask (1st track: 32)"
```

Comment:
if it is a bitmask, it should he a type of `b1`s.

Reply:
Thanks, I will look in to that.

Reply:
I’m sorry, but what? I just scoured the [KSY reference](http://doc.kaitai.io/ksy_reference.html#primitive-data-types) and can find no mention of single bit values or the `b` type indicator.

Reply:
I'll skip this so far. There are *way* too many obstacles on that path — i.e. alignment issues, eventual performance issues (allocating an ArrayList of booleans vs just a single 16-bit integer), etc.